include/fmt/dyn-args.h

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2020 Vladimir Solontsov
+// SPDX-License-Identifier: MIT Licence
+
+#ifndef FMT_DYN_ARGS_H_
+#define FMT_DYN_ARGS_H_
+
+#include <forward_list>
+#include <functional>
+#include <vector>
+
+#include "core.h"
+
+#if (defined(FMT_HAS_VARIANT) || __cplusplus >= 201703L)
+#  include <variant>
+#  ifndef FMT_HAS_VARIANT
+#    define FMT_HAS_VARIANT
+#  endif
+#endif
+
+FMT_BEGIN_NAMESPACE
+
+namespace internal {
+
+template <typename T, typename Char> struct is_string_view : std::false_type {};
+
+template <typename Char>
+struct is_string_view<basic_string_view<Char>, Char> : std::true_type {};
+
+#ifdef FMT_USE_STRING_VIEW
+template <typename Traits, typename Char>
+struct is_string_view<std::basic_string_view<Char, Traits>, Char>
+    : std::true_type {};
+#endif
+
+#ifdef FMT_USE_EXPERIMENTAL_STRING_VIEW
+template <typename Traits, typename Char>
+struct is_string_view<std::experimental::basic_string_view<Char, Traits>, Char>
+    : std::true_type {};
+#endif
+
+template <typename T> struct is_ref_wrapper : std::false_type {};
+
+template <typename T>
+struct is_ref_wrapper<std::reference_wrapper<T>> : std::true_type {};
+
+template <typename T, typename Context> struct need_dyn_copy {
+  using mapped_type = mapped_type_constant<T, Context>;
+  static_assert(mapped_type::value != internal::type::named_arg_type,
+                "Bug indicator. Named arguments must be processed separately");
+
+  using type = std::integral_constant<
+      bool, !(is_ref_wrapper<T>::value ||
+              is_string_view<T, typename Context::char_type>::value ||
+              (mapped_type::value != internal::type::cstring_type &&
+               mapped_type::value != internal::type::custom_type &&
+               mapped_type::value != internal::type::string_type))>;
+};
+
+template <typename T, typename Context>
+using need_dyn_copy_t = typename need_dyn_copy<T, Context>::type;
+
+template <typename T, typename StorageValue>
+const T& get(const StorageValue& v) {
+  return v;
+}
+
+}  // namespace internal
+
+/**
+  \rst
+  A dynamic version of `fmt::format_arg_store<>`.
+  It's equipped with a storage to potentially temporary objects which lifetime
+  could be shorter than the format arguments object.
+
+  It can be implicitly converted into `~fmt::basic_format_args` for passing
+  into type-erased formatting functions such as `~fmt::vformat`.
+  \endrst
+ */
+template <typename Context, typename... Args>
```

Comment:
The main point of dynamic argument store is that argument types may not be known at compile time, so this shouldn't be parameterized on `Args`. Custom args should be type erased at the time of insertion.

Also I don't think it's worth adding a new header just for this class. Please merge into `fmt/format.h`.

Reply:
I head to mention this in the comment -- it's required to support custom types.
But you're right, I can avoid it! 

Re separate header -- I didn't want to add a dependency from std::variant<> into the core.h.
But removing `Args` variant dependency will go away.

Reply:
Done